### PR TITLE
Wait for all files to be indexed in Azul before sending deletion (fixes #324)

### LINKED ISSUE #324: Race in data browser test may prevent all bundles from being deleted
The race occurs because [`_assert_data_browser_bundles()`](https://github.com/HumanCellAtlas/dcp/blob/2ab76d1ef70e48c438924670bae60319978f701d/tests/dataset_runner.py#L223) Waits until **at least one** file from each bundle is present, but not for **all** files from both bundles to be present. This means that some files may not be indexed before the deletion notification is sent. Thus these files will still be present in the index and Azul's cleanup assertion will fail.

This is what's responsible for this failure https://allspark.dev.data.humancellatlas.org/HumanCellAtlas/dcp/-/jobs/23210

The planned solution is to count the number of files referenced **per** bundles on a run that did not clean up. This number will be hard coded and `_assert_data_browser_bundles()` will be amended to check that the number of files referenced **per** bundles is the same.

### PR DESCRIPTION
**Pull request by jessebrennan**

### CODE REVIEW

File: tests/dataset_runner.py [hunk 1]
```diff
@@ -221,20 +221,27 @@ def assert_data_browser_bundles(self):
         ).to_return_value(value=True)
 
     def _assert_data_browser_bundles(self, project_shortname):
+        # Numbers here correspond to references to files for each bundle (AKA contributions). We must wait for all files
+        # references to be indexed before we continue to cleanup in order to avoid potential races.
+        project_to_expected_references = {'Smart-seq2': 22, '10x': 19}
```
Comment (hannes-ucsc) [inline tests/dataset_runner.py]:
```suggestion
        num_contributions_by_project = {'Smart-seq2': 22, '10x': 19}
```

File: tests/dataset_runner.py [hunk 2]
```diff
@@ -221,20 +221,27 @@ def assert_data_browser_bundles(self):
...
+        expected_bundle_uuids = set(self.primary_bundle_uuids).union(self.secondary_bundle_uuids)
+        files = self.azul_agent.get_entities_by_project('files', project_shortname)
+        bundle_uuids = {bundle['bundleUuid'] for file in files for bundle in file['bundles']}
+        project_shortnames = {project_short_name for file in files for project in file['projects']
+                              for project_short_name in project['projectShortname']}
+        num_files = sum(1 for f in files for _ in f['bundles'])
```
Comment (hannes-ucsc) [inline tests/dataset_runner.py]:
```suggestion
        num_files = sum(1 for f in files for b in f['bundles'])
```

File: tests/dataset_runner.py (hunk 1 above)
Comment (jessebrennan) [reply tests/dataset_runner.py]: Addressed

File: tests/dataset_runner.py (hunk 2 above)
Comment (jessebrennan) [reply tests/dataset_runner.py]: Addressed.

File: tests/dataset_runner.py [hunk 3]
```diff
@@ -369,20 +369,35 @@ def assert_data_browser_bundles(self):
         ).to_return_value(value=True)
 
     def _assert_data_browser_bundles(self, project_shortname):
+        # Numbers here correspond to references to files for each bundle (AKA contributions). We must wait for all files
+        # references to be indexed before we continue to cleanup in order to avoid potential races.
+        # A file is ref is a contribution if it appears in the bundle with "indexed": false.
+        # For ss2 we have 2 fastqs in the primary bundle; these 2 and 18 other contributions from the secondary
+        # For optimus we have 3 fastqs in the primary bundle; these 3 and 7 other contributions from the secondary
+        # for a total of 22 and 13 contributions respectively
```
Comment (hannes-ucsc) [inline tests/dataset_runner.py]: Great description.

File: tests/dataset_runner.py [hunk 4]
```diff
@@ -369,20 +369,35 @@ def assert_data_browser_bundles(self):
...
+        num_contributions = sum(1 for f in files for b in f['bundles'])
         try:
-            expected_bundle_uuids = set(self.primary_bundle_uuids).union(self.secondary_bundle_uuids)
-            files = self.azul_agent.get_entities_by_project('files', project_shortname)
-            bundle_uuids = {bundle['bundleUuid'] for file in files for bundle in file['bundles']}
             assert bundle_uuids == expected_bundle_uuids
-            project_shortnames = {project_short_name for file in files for project in file['projects']
-                                  for project_short_name in project['projectShortname']}
             assert project_shortnames == {project_shortname}
+            assert num_contributions == num_contributions_by_project[self.dataset.name]
```
Comment (hannes-ucsc) [inline tests/dataset_runner.py]: If `num_contributions < num_contributions_by_project[self.dataset.name]` changes to `num_contributions > num_contributions_by_project[self.dataset.name]`, this would time out even though we know the `>` case should never occur and constitutes an error. I think you can simplify this by not doing the `try: assert except AssertionError` which smells funny anyways. Simply `if … < …: return False`; elif … == …: return True; else: assert False` or equivalent. The `bundle_uuids` and `project_shortnames` asserts would have to be adjusted similarly.

File: tests/dataset_runner.py [hunk 5]
```diff
@@ -369,20 +369,35 @@ def assert_data_browser_bundles(self):
...
+        # for a total of 22 and 13 contributions respectively
+        num_contributions_by_project = {'Smart-seq2': 22, 'optimus': 13}
+        expected_bundle_uuids = set(self.primary_bundle_uuids).union(self.secondary_bundle_uuids)
+        files = self.azul_agent.get_entities_by_project('files', project_shortname)
+        bundle_uuids = {bundle['bundleUuid'] for file in files for bundle in file['bundles']}
+        project_shortnames = {project_short_name
+                              for file in files
+                              for project in file['projects']
+                              for project_short_name in project['projectShortname']}
+        num_contributions = sum(1 for f in files for b in f['bundles'])
```
Comment (hannes-ucsc) [inline tests/dataset_runner.py]:
```suggestion
        num_contributions = sum(1 for f in files for b in f['bundles'] if b in expected_bundle_uuids)
```

and …

File: tests/dataset_runner.py [hunk 6]
```diff
@@ -369,20 +369,35 @@ def assert_data_browser_bundles(self):
...
         try:
-            expected_bundle_uuids = set(self.primary_bundle_uuids).union(self.secondary_bundle_uuids)
-            files = self.azul_agent.get_entities_by_project('files', project_shortname)
-            bundle_uuids = {bundle['bundleUuid'] for file in files for bundle in file['bundles']}
             assert bundle_uuids == expected_bundle_uuids
```
Comment (hannes-ucsc) [inline tests/dataset_runner.py]: … 

```suggestion
            assert bundle_uuids >= expected_bundle_uuids
```

File: tests/dataset_runner.py (hunk 5 above)
Comment (jessebrennan) [reply tests/dataset_runner.py]: Sorry, I should have made my thinking clear. I saw the assert for the number of bundle uuids and concluded that that makes the filtering obsolete. I see why that isn't quite the case though, now that I see your sugestion to allow for `>=` expected bundles.

File: tests/dataset_runner.py (hunk 4 above)
Comment (jessebrennan) [reply tests/dataset_runner.py]: It would not time out because I added another assert in the `finally` clause https://github.com/HumanCellAtlas/dcp/pull/326/files#diff-b4fb3043d409f4155bdd0e6697aeaf78R400 I can change the design though.

**Comment by jessebrennan** [comment]
@hannes-ucsc I took into account your feedback and tried a new approach. It's not quite what you suggested, but I think it meets the requirements in a clearer, simpler way. Let me know if you agree.

File: tests/dataset_runner.py [hunk 7]
```diff
@@ -369,20 +369,42 @@ def assert_data_browser_bundles(self):
...
+        num_contributions = sum(1 for f in files for b in f['bundles'] if b['bundleUuid'] in expected_bundle_uuids)
+
+        Progress.report(f"{len(bundle_uuids)}/{len(expected_bundle_uuids)} distinct bundles")
+        Progress.report(f"{num_contributions}/{num_contributions_by_project[self.dataset.name]}"
+                        f" distinct file contributions")
+
+        if num_contributions < num_contributions_by_project[self.dataset.name]:
+            return False
+        elif num_contributions > num_contributions_by_project[self.dataset.name]:
+            raise ValueError('More contributions than expected were found in the Azul index')
```
Comment (hannes-ucsc) [inline tests/dataset_runner.py]:
```suggestion
            raise AssertionError('More contributions than expected were found in the Azul index')
```

File: tests/dataset_runner.py [hunk 8]
```diff
@@ -369,20 +369,42 @@ def assert_data_browser_bundles(self):
...
+
+        if bundle_uuids < expected_bundle_uuids:
+            return False
+
+        if not len(project_shortnames):
```
Comment (hannes-ucsc) [inline tests/dataset_runner.py]:
```suggestion
        if project_shortnames:
            assert ...
        else:
            return False
```

And decide if you want to follow `return` with the technically redundant `else` or not. Right now you're mixing both which I find confusing. So either no `else` after raise/return or everything should go into one giant `if …: … elif …: … elif …: … else: return True`

Comment (jessebrennan) [reply tests/dataset_runner.py]: Your suggested change is definitely clearer. Now that I've made that change it doesn't seem to me like anything is redundant. I think writing it like it is is a lot clearer than a bunch of nested `else`s, i.e.

        if num_contributions < num_contributions_by_project[self.dataset.name]:
            return False
        elif num_contributions > num_contributions_by_project[self.dataset.name]:
            raise ValueError('More contributions than expected were found in the Azul index')
        else:
            if bundle_uuids < expected_bundle_uuids:
                return False
            else:
                if project_shortnames:
                    assert project_shortnames == {project_shortname}
                else:
                    return False
        return True

**Comment by hannes-ucsc** [comment]
@jessebrennan, please merge into master, then merge master to integration, then integration to staging. I think all those merges can be done independently.

DON'T FORGET TO REMOVE THE COMMIT LABELED `Allow branch testing (temporary, REVERT ME)`

**Comment by parthshahva** [comment]
Do not merge integration into staging. Please merge master into staging and
prod as well

**Comment by hannes-ucsc** [comment]
What @parthshahva said, @jessebrennan.

**Comment by jessebrennan** [comment]
It is done. The change should be merged into the integration, staging, and prod. Thanks for your patience.

**Comment by hannes-ucsc** [comment]
BTW, @parthshahva I think I understand why you don't us want to merge integration to staging: because there are some golden egg changes on staging that would go away if we merged integration into it. 

However, I think that's the wrong model. As a rule, there should never be permanent diffs between deployment branches. We follow that rule Azul and it works well. The only permanent difference between two deployment branches should be their name. For example, the integration branch should always be ahead of the staging branch but the diff between the two branches should be a different set of changes every week. No single commit—the golden egg—should be expected to always be part of that set for things to work as expected. That's just too brittle.

More importantly, the golden egg commit prevents natural progression. Consider the situation we have now. @jessebrennan merged (and had to merge) from master into all three deployment branches without even knowing if his changes break the test on two of these branches: staging and prod. If we avoided the dependency of the golden egg commit, changes could progress naturally through the deployments because no golden egg commit prevents merging from a lower deployment branch to a higher one. And, this can happen as part of the weekly promotions. The PR author is relieved of the responsibility to merge his/her changes into all three deployment branches.

TL;DR we're doing it all wrong

PS: As usual, I can only pontificate and don't have the time to actually make it happen.

**Comment by sampierson** [comment]
This sounds worthy of a discussion at the Tech Arch Meeting.
